pkg/apis/core/validation/shoot.go

```diff
@@ -913,6 +913,45 @@ func validateWorkerGroupAndControlPlaneKubernetesVersion(controlPlaneVersion, wo
 	return allErrs
 }
 
+// validateDNSCredentialsRef validates the DNS provider credentials reference and secret name for backward compatibility.
+func validateDNSCredentialsRef(provider core.DNSProvider, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// How to achieve backward compatibility between secretName and credentialsRef?
+	// - if secretName is set, credentialsRef must be set and refer the same secret
+	// - if secretName is not set, then credentialsRef could be either unset or set but refer to a non-secret resource
+	//
+	// After the sync in the strategy, we can have the following cases:
+	// - both secretName and credentialsRef are unset, which we respect here
+	// - both can be set but refer to different resources, which we forbid here
+	// - secretName can be unset only when workloadIdentity is used, which we respect here
+
+	if provider.CredentialsRef == nil {
+		if provider.SecretName != nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("secretName"), "must not be set when `credentialsRef` is not set"))
+		}
```

Comment:
Nesting can be reduced here
```suggestion
	if provider.CredentialsRef == nil && provider.SecretName != nil {
			allErrs = append(allErrs, field.Forbidden(fldPath.Child("secretName"), "must not be set when `credentialsRef` is not set"))
```

Reply:
But then the else block will be satisfied in different condition: 

```go
provider.CredentialsRef != nil || provider.SecretName == nil
```

which is not the same as `provider.CredentialsRef != nil`

Reply:
You are of course right! My bad.